backend/service/aws/cache.go

```diff
@@ -0,0 +1,170 @@
+package aws
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/kinesis"
+	"github.com/golang/protobuf/ptypes"
+	"go.uber.org/zap"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+const topologyObjectChanBufferSize = 5000
+const topologyLockId = 1
+
+func (c *client) CacheEnabled() bool {
+	return true
+}
+
+func (c *client) StartTopologyCaching(ctx context.Context) (<-chan *topologyv1.UpdateCacheRequest, error) {
+	if !c.topologyLock.TryAcquire(topologyLockId) {
+		return nil, errors.New("TopologyCaching is already in progress")
+	}
+
+	go func() {
+		<-ctx.Done()
+		close(c.topologyObjectChan)
+		c.topologyLock.Release(topologyLockId)
+	}()
+
+	c.processRegionTopologyObjects(ctx)
+	return c.topologyObjectChan, nil
+}
+
+// We process each resource and region individually, creating tickers for each.
+// If a single resource and region takes longer to process it will not block other resources.
+// Additionally this give us the flexibility to tune the frequency based on resource.
+func (c *client) processRegionTopologyObjects(ctx context.Context) {
+	for name, client := range c.clients {
+		c.log.Info("processing topology objects for region", zap.String("region", name))
+		go c.startTickerForCacheResource(ctx, time.Duration(time.Minute*5), client, c.processAllEC2Instances)
+		go c.startTickerForCacheResource(ctx, time.Duration(time.Minute*10), client, c.processAllAutoScalingGroups)
+		go c.startTickerForCacheResource(ctx, time.Duration(time.Minute*10), client, c.processAllKinesisStreams)
+	}
+}
+
+func (c *client) startTickerForCacheResource(ctx context.Context, duration time.Duration, client *regionalClient, resourceFunc func(context.Context, *regionalClient)) {
+	ticker := time.NewTicker(duration)
+
+	go func() {
+		<-ctx.Done()
+		ticker.Stop()
+	}()
+
+	for ; true; <-ticker.C {
+		resourceFunc(ctx, client)
+	}
+}
```

Comment:
ticker.Stop() does not close the channel so this will result in a deadlock, need to rethink.

example: https://play.golang.org/p/goI6_5gcBma

Reply:
this works nicely: https://play.golang.org/p/NMjqyInnB1E

Reply:
ah I see now :( alright thanks

Reply:
Unable to use `time.Tick()` directly but thats alright I changed it to use `time.NewTicker` instead.

```
service/aws/cache.go:53:12: SA1015: using time.Tick leaks the underlying ticker, consider using it only in endless functions, tests and the main package, and use time.NewTicker here (staticcheck)
	ticker := time.Tick(duration)
```